### PR TITLE
Stft add named windows

### LINKED ISSUE #88919: Make 'hann' the default window in torch.stft
### 🚀 The feature, motivation and pitch

The current default window used on Torch's Short Time Fourier Transform (STFT) method is the rectangular window.
This window has major drawbacks and is known to produce artifacts in the output (see [here](https://www.audiolabs-erlangen.de/resources/MIR/FMP/C2/C2_STFT-Window.html) for details.

Both [Scipy's implementation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html) and [Librosa's](https://librosa.org/doc/main/generated/librosa.stft.html) use the Hann window by default. Furthermore, both libraries allow for a string name to be used, abstracting the creation of the window itself.

I propose the following changes to `torch.stft`:
* Make the Hann window the default on function `torch.stft`
* Modify the interface of torch.stft to allow a string to be provided to the `window` parameter, and include common windows such as `torch.hann_window`, `torch.blackman_window` and `torch.hamming_window` 

I would be happy to work on this myself. I propose to and test the code below to `torch.stft`, and add relevant documentation as well:

```python    if isinstance(window, str):
        windows = {
            "hann": torch.hann_window,
            "hamming": torch.hamming_window,
            "blackman": torch.blackman_window
        }
        if window not in windows.keys():
            raise ValueError(f"""{window} is not a valid window name.
            			Available windows are {windows.keys()}""")
        
        else:
            win_length = n_fft if win_length is None else win_length
            window = windows[window](win_length, device=input.device)
```

I hope this will be a positive contribution to the Pytorch community. Please tell me if you prefer for this to be done in a different way.

Kind regards,

Eric

### Alternatives

_No response_

### Additional context

_No response_

cc @mruberry @peterbell10

### PR DESCRIPTION
**Pull request by egrinstein**
Fixes #88919

As discussed in issue #88919, here is the pull request that adds the option to use a named window for the STFT. This PR

* adds the functionality directly to the Python code on the `functional.py` module. the `bartlett`, `blackman`, `hamming`, `hann`, and `kaiser` windows are supported.
* modifies the documentation to describe this new option and to mention that not specifying a window with throw an error in version 1.15.0 (is this the correct version?).
* adds a warning if a window is not specified
* adds a unit test in `test/test_spectral_ops.py`. The test verifies that using a window string produces the same result as providing a precomputed Tensor window of the same type.

I did my best to follow the contribution guidelines, but as this is my first PR, thank you for pointing out any mistakes I might have made.

@mruberry, should I create these two new issues:
* the first to be solved before 1.15.0, which throws an error if the window is not specified
* the second to be solved before 1.16.0, which changes the default window to `hann`

Best wishes,

Eric

@mruberry @peterbell10

### CODE REVIEW

File: torch/functional.py [hunk 1]
```diff
@@ -596,7 +600,7 @@ def stft(input: Tensor, n_fft: int, hop_length: Optional[int] = None,
             frames. Default: ``None`` (treated as equal to ``floor(n_fft / 4)``)
         win_length (int, optional): the size of window frame and STFT filter.
             Default: ``None``  (treated as equal to :attr:`n_fft`)
-        window (Tensor, optional): the optional window function.
+        window (Tensor, str, optional): the optional window function.
```
Comment (mruberry) [inline torch/functional.py:603]: This is misleading (it was misleading before, too) because this PR will allow users to specify a window function to generate the window OR to provide a tensor to be used as the window directly

File: torch/functional.py [hunk 2]
```diff
@@ -621,10 +625,39 @@ def stft(input: Tensor, n_fft: int, hop_length: Optional[int] = None,
                Note that calling :func:`torch.view_as_real` on the output will
                recover the deprecated output format.
 
+            .. deprecated:: 1.15.0
```
Comment (mruberry) [inline torch/functional.py:628]: Let's move the deprecation and warning to a separate PR and have this PR just add the new functionality

File: test/test_spectral_ops.py [hunk 3]
```diff
@@ -1093,6 +1092,32 @@ def test_complex_stft_definition(self, device, dtype):
             actual = torch.stft(*args, window=window, center=False)
             self.assertEqual(actual, expected)
 
+    @onlyNativeDeviceTypes
+    @skipCPUIfNoFFT
+    @dtypes(torch.double, torch.cdouble)
+    def test_stft_named_windows(self, device, dtype):
```
Comment (mruberry) [inline test/test_spectral_ops.py:1098]: Instead of this test, could the sample inputs for stft be extended?

**Comment by mruberry** [comment]
@egrinstein Thanks for this PR!

As I understand this issue better, I have some additional thoughts and questions.

In particular, I don't love specifying a window generator using a string. Why not just ask users to call the window function themselves and pass the generated tensor to `stft` and `istft`? (Should this PR update `istft` at the same time as `stft`?) The window functions take additional arguments (like `periodic`), and so specifying them as a string kinda underspecifies how they'll be used. It also creates two ways to do the same thing (having the user create the window themselves or pass a string describing how to construct the window). 

What are your thoughts, @peterbell10?

So I could see warning when a user fails to specify `window` explicitly, and that warning could provide an explanation of the current behavior and our recommendation (call this windowing function beforehand). 

We would also have to audit all our `stft` (and `istft`?) call sites to ensure PyTorch doesn't throw the warning. Do we ever call `stft` without explicitly specifying the window?

**Comment by egrinstein** [comment]
Hi @mruberry, thank you very much for the review!

You make a compelling point; I don't love using a string to generate the window either. I have initially done it this way because it is how it is in Scipy and Librosa. Also, thank you for pointing about the istft, I had forgotten about it.

However, since many users come from the above libraries, they may be inadvertedly using the rectangular window (I confess I was), which has no use case I can think of. I think that the `hann` window suits most users, so it would be good to make it the default behaviour.

Based on this, I would propose to:

1. Abandon this PR
2. Create a PR raising a warning on `stft` and `istft` when a window is not provided.
3. Wait 180 days
4. Create a PR which throws an error if a window is not provided
5. Wait 180 days
6. Create a PR to create a `hann` window on `stft` and `istft` when `window=None` instead of using a rectangular window. This would be leaner than implementing the window strings, as we would only be maintaining a single window. Alternatively, we could make the `window` parameter required.

What do you think?

**Comment by mruberry** [comment]
> Hi @mruberry, thank you very much for the review!
> 
> You make a compelling point; I don't love using a string to generate the window either. I have initially done it this way because it is how it is in Scipy and Librosa. Also, thank you for pointing about the istft, I had forgotten about it.
> 
> However, since many users come from the above libraries, they may be inadvertedly using the rectangular window (I confess I was), which has no use case I can think of. I think that the `hann` window suits most users, so it would be good to make it the default behaviour.
> 
> Based on this, I would propose to:
> 
> 1. Abandon this PR
> 2. Create a PR raising a warning on `stft` and `istft` when a window is not provided.
> 3. Wait 180 days
> 4. Create a PR which throws an error if a window is not provided
> 5. Wait 180 days
> 6. Create a PR to create a `hann` window on `stft` and `istft` when `window=None` instead of using a rectangular window. This would be leaner than implementing the window strings, as we would only be maintaining a single window. Alternatively, we could make the `window` parameter required.
> 
> What do you think?

I like throwing a warning for `stft` and `istft` if a window is not provided, and making it an error to do so sounds like the right approach. How about we re-evaluate then if we want to provide a new default? The warning should inform users about why the warning is occurring, how to get the current default behavior without a warning, and what our recommendation is, instead. The PR that throws the warning should also audit our `stft` and `istft` callsites so people don't get warnings when calling PyTorch operations.

So basically I agree up through part (5) of your proposal and then think we can reconsider (6) then. Sound good, @egrinstein? I think this warning will really help `stft` and `istft` users.

**Comment by egrinstein** [comment]
Hi @mruberry,

I like the plan.

I will start working on a new PR which adds a warning on STFT and ISTFT, find where these functions are called, and filter the warning on them.
